plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java

```diff
@@ -3068,6 +3095,36 @@ public MaterializedViewFreshness getMaterializedViewFreshness(ConnectorSession s
         return hiveMaterializedViewMetadata.refreshMaterializedView(session, name);
     }
 
+    public Optional<CatalogSchemaTableName> redirectTable(ConnectorSession session, SchemaTableName tableName)
+    {
+        if (isRedirectIcebergTables(session) && filterSchema(tableName.getSchemaName()) && isExistingIcebergTable(session, tableName)) {
+            return Optional.of(new CatalogSchemaTableName(getTargetIcebergCatalog(session), tableName));
+        }
+        return Optional.empty();
+    }
+
+    private boolean isExistingIcebergTable(ConnectorSession session, SchemaTableName schemaTableName)
+    {
+        String tableName = schemaTableName.getTableName();
+        Matcher icebergNameMatch = ICEBERG_TABLE_NAME_PATTERN.matcher(tableName);
+        if (icebergNameMatch.matches()) {
+            tableName = icebergNameMatch.group("table");
+        }
+        Optional<Table> table = metastore.getTable(new HiveIdentity(session), schemaTableName.getSchemaName(), tableName);
+        if (table.isEmpty()) {
+            return false;
+        }
+        return ICEBERG_TABLE_TYPE_VALUE.equalsIgnoreCase(table.get().getParameters().get(TABLE_TYPE_PROP));
+    }
+
+    private void rejectRedirectedTables(ConnectorSession session, SchemaTableName tableName)
+    {
+        if (redirectTable(session, tableName).isPresent()) {
+            throw new TrinoException(NOT_SUPPORTED, "Hive Connector doesn't support modification operations (write, DDL, comment, statistics collection, set authorization) " +
+                    "on Iceberg tables when redirection is enabled");
```

Comment:
Do you mean that redirects on modification operations are a bad idea?
then why do we support them at all?

if they are a good idea in general, why would we want to opt out here?

Reply:
@findepi

> Do you mean that redirects on modification operations are a bad idea?

https://github.com/trinodb/trino/pull/7606#issuecomment-843542294

does it look better with `Fail modifications for redirected tables in engine` commit? If so, I can put it as the first commit. Note that Hive connector check is still required because of procedures.

Reply:
> does it look better with `Fail modifications for redirected tables in engine` commit?

I am not very fluent with this code, but if this implements @electrum 's thinking https://github.com/trinodb/trino/pull/7606#issuecomment-843542294, it should go in separate PR and to be debated separately. There hopefully is nothing special about redirects in Iceberg or Hive connector to warrant DDL-specific checks in the connector code (except, maybe, procedures -- which ones?)

Reply:
makes sense to do it in a separate PR, was trying to gather feedback if exceptions thrown in hive redirection tests feel more natural with this change. and yes, that commit doesn't have anything specific to hive connector. 

hive connector checks become sort of "illegal state checks" in non-procedure calls, but wouldn't hurt to still keep proper error message there.

Reply:
> but wouldn't hurt to still keep proper error message there.

except that it suggests -- to the future reader -- that we're coding some connector-specific behavior.

(and I do care about future readers quite a lot:)

Reply:
Good point. How about changing all connector checks to `checkState`, and adding special handling for procedures that throw TrinoException?

Reply:
or TrinoException here, but with a comment. up to you